### ISSUE TITLE
Contract function that references `msg.sender` fails when called from within another contract

### ISSUE THREAD

**Issue by Grety** [ethereum/solidity#1682]
Is it supposed to be like this and I am just missing some docs or is it a bug?

After getting stuck with own implementation I started searching web. The closest MCVE is this [ethereum.stackexchange.com answer](http://ethereum.stackexchange.com/a/2831/4853). 

I simplified Solidity code from that post a little. There is no need in base contract `mortal` and I pass the address of `Name` instance (`watch_addr` from the post) for the sake of changeability.
```
pragma solidity ^0.4.9;

contract Name {

    mapping(address=>string) public text;
    string public test;

    function register(string _text){
        text[msg.sender]    = _text;
    }

    function getName() constant returns (string) {
    	return text[msg.sender];
    }
}

contract Proxy {

    address user_addr   = msg.sender;

    function register(string _text, address watch_addr){
        Name name = Name(watch_addr);
        name.register(_text);
    }
}
```

When used via `Proxy` the method call `register` silently throws internally and has no affect on the storage.
```
...
yield name.register('test name');
console.log(`registered name is ${name.getName()}`);// registered name is test name
yield proxy.register('proxy overridden name', addresses.Name);
console.log(`registered name is ${name.getName()}`);// registered name is test name
...
```
However, if I change the text of contract function `Name.register` and `Name.getName`  to work with a static address (it works well with just `0x001`) instead of `msg.sender` the last line in js code prints `registered name is proxy overriden name`.

**Environment**: Geth/v1.5.5-stable-ff07d548/linux/go1.7.3, js code runs in Node v7.4.0, compiler: npm/solc/v0.4.9

Thanks for looking into this!

**Comment by axic** [comment]
I'm not sure what setup you are using, but it works well with [remix](https://remix.ethereum.org/). Also note that `msg.sender` will be the address of the `Proxy` instance.

**Comment by chriseth** [comment]
The only reason I see why this would throw is that you pass the wrong address. Note that every function call checks that there actually is code at the other end. If you pass a wrong address, it will throw.

**Comment by Grety** [comment]
Sorry for troubling. Eventually it turned out that the reason was Out of gas for transaction. I was not sending any additional gas, just default value. So it crashed at various places on heavy operations. But still [as far as I understood](https://github.com/ethereum/go-ethereum/issues/3163) gas estimation doesn't work for contract methods... At least I failed to set it up correctly. Now I do estimation but it always estimates to consume all sent gas. Which is not the case if I send real transaction. So I just ran several tests and now send excessive amounts of gas for my contract calls
